libs/ml/include/ml/layers/fully_connected.hpp

```diff
@@ -189,22 +259,66 @@ class FullyConnected : public SubGraph<T>
             inputs.front()->shape(inputs.front()->shape().size() - 1)};
   }
 
+  math::SizeVector ComputeBatchOutputShape(
+      std::vector<math::SizeVector> const &input_shapes) override
+  {
+    if (time_distributed_)
+    {
+      assert((this->in_size_ == AUTODETECT_INPUT_SHAPE) ||
+             (input_shapes.front().at(0) == in_size_));
+
+      this->SetBatchInputShapes(input_shapes);
+      if (input_shapes.front().size() == 3)
+      {
+        this->SetBatchOutputShape({this->out_size_, input_shapes.front().at(1), 1});
+      }
+      else
+      {
+        this->SetBatchOutputShape({this->out_size_, 1, 1});
+      }
+      return this->batch_output_shape_;
+    }
+
+    this->SetBatchInputShapes(input_shapes);
+    this->SetBatchOutputShape({this->out_size_, 1});
+    return this->batch_output_shape_;
+  }
+
   static constexpr OpType OpCode()
   {
     return fetch::ml::OpType::LAYER_FULLY_CONNECTED;
   }
 
   static constexpr char const *DESCRIPTOR = "FullyConnected";
 
+  OpType OperationType() const override  // TODO(ML-466) : move implementation to .cpp
+  {
+    return this->OpCode();
+  }
+  char const *Descriptor() const override  // TODO(ML-466) : move implementation to .cpp
+  {
+    return DESCRIPTOR;
+  }
+
+  static constexpr SizeType AUTODETECT_INPUT_SHAPE = 0;
+  static constexpr bool     TIME_DISTRIBUTED       = true;
+
 private:
+  // Saveable params
   SizeType in_size_          = fetch::math::numeric_max<SizeType>();
   SizeType out_size_         = fetch::math::numeric_max<SizeType>();
   bool     time_distributed_ = false;
 
-  void Initialise(TensorType &weights, WeightsInit init_mode)
-  {
-    fetch::ml::ops::Weights<TensorType>::Initialise(weights, in_size_, out_size_, init_mode);
-  }
+  // Non-saveable params
+  bool                          is_initialised_ = false;
```

Comment:
all these non-saveable params leave me slightly nervous about whether dense layer serialisation->deserialisation will ALWAYS result in the same behaviour.

For example. is_initialised_ can be true for a dense layer. Serialise and deserialise the layer and the new dense layer has is_initialised_ = false. Couldn't that be a cause of differing behaviour in the layers?

Reply:
the safe (but also less efficient) thing to do is serialised all the parameters - then behaviour is guaranteed; but you're right that there are definitely some parameters that are just effectively internal caching and don't need to be serialised.

Reply:
I agree. I have removed all fields possible, leaving only what are really necessary to keep, and serialised them.